pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gabagecollector
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const ResourceResyncTime = 30 * time.Second
+
+type GarbageCollector struct {
+	restMapper         meta.RESTMapper
+	clientPool         dynamic.ClientPool
+	dirty_queue        *workqueue.Type
+	scannerStores      []cache.Store
+	scannerControllers []*framework.Controller
+}
+
+func NewGarbageCollector(clientPool dynamic.ClientPool, resources []unversioned.GroupVersionResource) *GarbageCollector {
+	gc := &GarbageCollector{
+		clientPool:  clientPool,
+		dirty_queue: workqueue.New(),
+		// TODO: should use a dynamic RESTMapper built from the discovery results.
+		restMapper: registered.RESTMapper(),
+	}
+	for _, resource := range resources {
```

Comment:
We should be very careful about adding things to the discovery docs... they're unversioned and we don't get a second change on additions.

The idea of a canonical resource could be useful, though it raises several questions:
* do we always have enough information to know what the canonical resource is?
* does that imply the canonical resource is always the superset?
* as versions progress, would the canonical resource change to point to the newest version?
* when the old versions aren't available via the API any longer, references to them would not be resolveable, even if the underlying object was still available (e.g. a reference to minion/v1beta3 wouldn't be traceable to node/v1 since minion/v1beta3 wouldn't be in the discovery doc any longer). Does that imply the relationship should go the other way (node/v1 could indicate it was a canonical resource of minion/v1beta3)?

For `ManagedByGC`, making resources declare controller-specific attributes seems inverted to me.